libs/langgraph/src/pregel/runner.ts

```diff
@@ -131,6 +148,79 @@ export class PregelRunner {
     }
   }
 
+  /**
+   * Initializes the current AbortSignals for the PregelRunner, handling the various ways that
+   * AbortSignals must be chained together so that the PregelLoop can be interrupted if necessary
+   * while still allowing nodes to gracefully exit.
+   *
+   * This method must only be called once per PregelRunner#tick. It has the side effect of updating
+   * the PregelLoop#config with the new AbortSignals so they may be propagated correctly to future
+   * ticks and subgraph calls.
+   *
+   * @param options - Options for the initialization.
+   * @returns The current abort signals.
+   * @internal
+   */
+  private _initializeAbortSignals({
+    exceptionSignalController,
+    timeout,
+    signal,
+  }: {
+    exceptionSignalController: AbortController;
+    timeout?: number;
+    signal?: AbortSignal;
+  }): PregelAbortSignals {
+    const previousSignals: PregelAbortSignals =
+      (this.loop.config.configurable?.[
+        CONFIG_KEY_ABORT_SIGNALS
+      ] as PregelAbortSignals) ?? {};
+
+    // This is true when a node calls a subgraph and, rather than forwarding its own AbortSignal,
+    // it creates a new AbortSignal and passes that along instead.
+    const subgraphCalledWithSignalCreatedByNode =
+      signal &&
+      previousSignals.composedAbortSignal &&
+      signal !== previousSignals.composedAbortSignal;
+
+    const externalAbortSignal = subgraphCalledWithSignalCreatedByNode
+      ? // Chain the signals here to make sure that the subgraph receives the external abort signal in
+        // addition to the signal created by the node.
+        combineAbortSignals(previousSignals.externalAbortSignal!, signal!)
+      : // Otherwise, just keep using the external abort signal, or initialize it if it hasn't been
+        // assigned yet
+        previousSignals.externalAbortSignal ?? signal;
+
+    const errorAbortSignal = previousSignals.errorAbortSignal
+      ? combineAbortSignals(
+          previousSignals.errorAbortSignal!,
+          exceptionSignalController.signal
+        )
+      : exceptionSignalController.signal;
```

Comment:
Probably also worth a comment - chaining here rather than always using a fresh one handles the case where a subgraph is called in a parallel branch to some other node in the parent graph. Honestly I'm not even sure that we support that, but if we do, we propagate the `AbortSignal` correctly in that case now.